lib/puppetlabs_spec_helper/tasks/module_inspector.rb

```diff
@@ -0,0 +1,102 @@
+require 'set'
+require 'puppet'
+
+module PuppetlabsSpecHelper; end
+module PuppetlabsSpecHelper::Tasks; end
+
+class PuppetlabsSpecHelper::Tasks::ModuleInspector
+  attr_reader :path
+
+  def initialize(path = Dir.pwd)
+    @path = path
+  end
+
+  def run
+    {
+      host_class:   host_classes,
+      defined_type: defined_types,
+    }
+  end
+
+  private
+
+  def manifests
+    return [] unless File.directory?(File.join(path, 'manifests'))
+
+    Dir[File.join(path, 'manifests', '**', '*.pp')]
+  end
+
+  def manifest_definitions
+    @manifest_definitions ||= manifests.map { |r|
+      PuppetlabsSpecHelper::Tasks::ManifestInspector.new(r).definitions
+    }.flatten
+  end
+
+  def host_classes
+    remove_type_key(manifest_definitions.select { |r| r[:type] == :host_class })
+  end
+
+  def defined_types
+    remove_type_key(manifest_definitions.select { |r| r[:type] == :defined_type })
+  end
+
+  def remove_type_key(hash)
+    hash.map { |r| r.reject { |k, _| k == :type } }
+  end
+end
+
+class PuppetlabsSpecHelper::Tasks::ManifestInspector
+  attr_reader :path
+
+  DEFINITION_CLASSES = Set[
+    Puppet::Pops::Model::HostClassDefinition,
+    Puppet::Pops::Model::ResourceTypeDefinition,
+  ]
+
+  def initialize(path)
+    @path = path
+  end
+
+  def definitions
+    ast.definitions.find_all { |r| DEFINITION_CLASSES.include?(r.class) }.map do |obj|
```

Comment:
I assume if something "goes wrong" here the error is being surfaced via stderr and exit codes.  Is there a case to capture this and output in structured format? e.g.

``` json
{
  "error":  "THE WORLD IS CRASHING!"
  "file": "/some/path"
}
```

Should we return partial results?
Currently no tests for unhappy path of invalid manifests etc.